bionic/provider.py

```diff
@@ -156,24 +155,31 @@ def __init__(self, names, protocols, docs):
                 can_memoize=True,
                 changes_per_run=False,
             ),
-            is_mutable=True,
         )
 
-        self.clear_cases()
+        self.key_space = CaseKeySpace()
+        self._has_any_values = False
+        self._value_tuples_by_case_key = {}
+        self._token_tuples_by_case_key = {}
 
-    def copy(self):
+    def copy_and_add_case(self, case_key, values):
```

Comment:
now that all the providers are immutable, you can drop `copy_and_` since we know it will return a new object (much like how other immutable objects work in bionic, like flow state).

Reply:
Yeah... I guess that's consistent with how FlowState works. I was more explicit here because, unlike FlowState, providers don't have any other methods that behave like this, but I guess `add_case` is still pretty clear.